odoo/addons/base/models/ir_module.py

```diff
@@ -650,13 +650,23 @@ def button_upgrade(self):
                 raise UserError(_("Can not upgrade module '%s'. It is not installed.") % (module.name,))
             if self.get_module_info(module.name).get("installable", True):
                 self.check_external_dependencies(module.name, 'to upgrade')
-            for dep in Dependency.search([('name', '=', module.name)]):
-                if (
-                    dep.module_id.state == 'installed'
-                    and dep.module_id not in todo
-                    and dep.module_id.name != 'studio_customization'
-                ):
-                    todo.append(dep.module_id)
+            if module.name == 'base':
+                # If an installed module is only present in the dependency graph through
+                # a new, uninstalled dependency, it will not have been selected yet.
+                # An update of 'base' should also update these modules, and as a consequence,
+                # install the new dependency.
+                dependent_modules = self.search(
+                    [('state', '=', 'installed'), ('name', '!=', 'studio_customization')])
```

Comment:
My suggestion was to move the `if` statement above the `while` loop itself.
If you upgrade `base`, then `todo` should be all the installed modules, otherwise you determine `todo` with this loop.
Don't you think so?

Reply:
I wasn't quite sure what you meant. How about now?